In [1]:
import pandas as pd 

In [3]:
df = pd.read_csv("../data/telecom_churn_raw.csv")


In [5]:
df.head()

Customer ID  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV    Male   78       No            Yes      No         No   
1  7495-OOKFY  Female   74       No            Yes     Yes        Yes   
2  1658-BYGOY    Male   71       No            Yes      No        Yes   
3  4598-XLKNJ  Female   78       No            Yes     Yes        Yes   
4  4846-WHAFZ  Female   80       No            Yes     Yes        Yes   

   Number of Dependents        Country       State  ...  \
0                     0  United States  California  ...   
1                     1  United States  California  ...   
2                     3  United States  California  ...   
3                     1  United States  California  ...   
4                     1  United States  California  ...   

  Total Extra Data Charges  Total Long Distance Charges  Total Revenue  \
0                       20                         0.00          59.65   
1                        0                       390.80        1024.10   
2                        0                       203.94        1910.88   
3                        0                       494.00        2995.07   
4                        0                       234.21        3102.36   

   Satisfaction Score  Customer Status Churn Label Churn Score  CLTV  \
0                   3          Churned         Yes          91  5433   
1                   3          Churned         Yes          69  5302   
2                   2          Churned         Yes          81  3179   
3                   2          Churned         Yes          88  5337   
4                   2          Churned         Yes          67  2793   

    Churn Category                  Churn Reason  
0       Competitor  Competitor offered more data  
1       Competitor  Competitor made better offer  
2       Competitor  Competitor made better offer  
3  Dissatisfaction     Limited range of services  
4            Price            Extra data charges  

[5 rows x 50 columns]

In [9]:

df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
)


In [11]:
df.columns

Index(['customer_id', 'gender', 'age', 'under_30', 'senior_citizen', 'married',
       'dependents', 'number_of_dependents', 'country', 'state', 'city',
       'zip_code', 'latitude', 'longitude', 'population', 'quarter',
       'referred_a_friend', 'number_of_referrals', 'tenure_in_months', 'offer',
       'phone_service', 'avg_monthly_long_distance_charges', 'multiple_lines',
       'internet_service', 'internet_type', 'avg_monthly_gb_download',
       'online_security', 'online_backup', 'device_protection_plan',
       'premium_tech_support', 'streaming_tv', 'streaming_movies',
       'streaming_music', 'unlimited_data', 'contract', 'paperless_billing',
       'payment_method', 'monthly_charge', 'total_charges', 'total_refunds',
       'total_extra_data_charges', 'total_long_distance_charges',
       'total_revenue', 'satisfaction_score', 'customer_status', 'churn_label',
       'churn_score', 'cltv', 'churn_category', 'churn_reason'],
      dtype='object')




- **Target variable:** churn_label  
- **Meaning:**
  - Yes → Customer has churned  
  - No → Customer is retained



In [18]:
churn_rate = df['churn_label'].value_counts(normalize=True) * 100
churn_rate


churn_label
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

### Overall Churn Rate

- Approximately **26.5%** of customers have churned.
- This indicates a significant customer retention challenge for the telecom company.



In [22]:
churn_by_contract = (
    df.groupby('contract')['churn_label']
      .value_counts(normalize=True)
      .rename('percentage')
      .mul(100)
      .reset_index()
)

churn_by_contract


contract churn_label  percentage
0  Month-to-Month          No   54.155125
1  Month-to-Month         Yes   45.844875
2        One Year          No   89.290323
3        One Year         Yes   10.709677
4        Two Year          No   97.450876
5        Two Year         Yes    2.549124

### Churn by Contract Type

- Customers on **month-to-month contracts** show significantly higher churn.
- Long-term contracts (one year, two year) have much lower churn rates.
- This suggests that lack of long-term commitment increases churn risk.


In [25]:

df['tenure_group'] = pd.cut(
    df['tenure_in_months'],
    bins=[0, 12, 24, 48, 72],
    labels=['0-12 months', '13-24 months', '25-48 months', '49-72 months']
)


In [27]:
churn_by_tenure = (
    df.groupby('tenure_group')['churn_label']
      .value_counts(normalize=True)
      .rename('percentage')
      .mul(100)
      .reset_index()
)

churn_by_tenure


C:\Users\gk413\AppData\Local\Temp\ipykernel_16156\2982309646.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('tenure_group')['churn_label']


tenure_group churn_label  percentage
0   0-12 months          No   52.561757
1   0-12 months         Yes   47.438243
2  13-24 months          No   71.289062
3  13-24 months         Yes   28.710938
4  25-48 months          No   79.611041
5  25-48 months         Yes   20.388959
6  49-72 months          No   90.486824
7  49-72 months         Yes    9.513176

### Churn by Tenure

- Customers with **short tenure (0–12 months)** show the highest churn.
- Churn decreases significantly as tenure increases.
- This indicates that early customer experience is critical for retention.


### Key Analytical Findings (Input for Generative AI)

- Overall churn rate: ~26.5%
- Month-to-month contract churn: ~45.8%
- One-year contract churn: ~10.7%
- Two-year contract churn: ~2.5%
- 0–12 months tenure churn: ~47.4%
- Churn decreases consistently as tenure increases


Based on the following churn analysis findings: - Overall churn rate is approximately 26.5%. - Customers on month-to-month contracts have around 45.8% churn. - One-year contract churn is around 10.7%. - Two-year contract churn is around 2.5%. - Customers in their first 12 months show around 47.4% churn. - Churn decreases significantly as tenure increases. Explain the main drivers of churn in simple business language and suggest 3 actionable recommendations for reducing churn. What is driving customer churn?
1. Short-term contracts = low commitment

Customers on month-to-month contracts are leaving at a very high rate (~46%).
This tells us they:

Are not emotionally or financially locked in

Can leave the moment they feel unhappy with price, service, or experience

Often treat the service as replaceable

👉 In contrast, customers on 1-year and 2-year contracts stay much longer because switching feels costly and inconvenient.

2. First-year experience is critical

Nearly half of customers churn within the first 12 months.

This usually means:

Onboarding is weak or confusing

Expectations set during sales don’t match reality

Early service issues (billing, network quality, customer support) are not resolved fast enough

👉 If customers don’t see value early, they exit before forming loyalty.

3. Tenure builds loyalty and habit

Churn drops sharply as tenure increases.

Long-tenure customers:

Are comfortable with the service

Have adapted their usage to the network

Face higher switching costs (time, effort, risk)

👉 This shows churn is mainly an early-stage problem, not a long-term one.

3 Actionable Recommendations to Reduce Churn
1. Convert month-to-month users to longer contracts

Action:

Offer discounts, extra data, or free add-ons for moving to 1-year plans

Run targeted campaigns only for month-to-month customers

Business impact:

Even small conversion to longer contracts can cut churn by 50–80% in that segment

2. Strengthen first-90-days customer experience

Action:

Proactive welcome calls/messages

Early check-ins at 30, 60, and 90 days

Fast-track issue resolution for new customers

Business impact:

Reduces early dissatisfaction

Builds trust before the “easy exit” mindset kicks in

3. Early churn-risk detection & intervention

Action:

Flag customers with:

Low tenure

Month-to-month contracts

Complaints or reduced usage

Offer retention incentives before they decide to leave

Business impact:

Cheaper than acquiring new customers

Improves customer lifetime value (CLV)

Executive Summary (1-line)

Churn is driven mainly by low commitment plans and poor early-stage experience; fixing the first year and locking customers into longer contracts will deliver the biggest churn reduction.

### Key Analytical Findings

- Overall churn rate: ~26.5%
- Month-to-month contract churn: ~45.8%
- One-year contract churn: ~10.7%
- Two-year contract churn: ~2.5%
- 0–12 months tenure churn: ~47.4%
- Churn decreases consistently as tenure increases


In [ ]:


The following findings were provided to a language model to generate executive-level insights and recommendations.


In [36]:
dashboard_cols = [
    'customer_id',
    'gender',
    'age',
    'senior_citizen',
    'married',
    'tenure_in_months',
    'contract',
    'payment_method',
    'monthly_charge',
    'total_charges',
    'satisfaction_score',
    'churn_label'
]

df_dashboard = df[dashboard_cols]


In [42]:
df_dashboard.to_csv("../data/telecom_churn_powerbi.csv", index=False)


In [44]:
df_dashboard.head()
df_dashboard.shape


(7043, 12)